docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+  - note that with `versionId` set to an empty string `''`, it will overwrite
+    the master version only (same as no options, but returns a version ID)
+
+#### DELETE
+
+- no options: original DELETE operation, will delete the master version
+- `versionId: <versionId>` delete a specific version
+
+Version-specific deletions involve a few steps. A deletion targeting the latest
+version of an object has to:
+
+- delete the specified version identified by `versionId`
+- replace the master version with a version that is a placeholder for deletion
+  - this version contains a special keyword, 'isPHD', to indicate the master
+  version was deleted and needs to be updated
+- initiate a repair operation to update the value of the master version:
+  - involves listing the versions of the object and get the latest
+  version to replace the place holder delete version
+  - if no more versions exists, metadata deletes the master version, removing
+  the key from metadata
+
+Note: all of this happens before responding to S3, and only when the metadata
+engine is instructed by S3 to delete a specific version or the master version.
+See section "Delete Markers" for a description of what happens when a delete
+object request is sent to the S3 API.
+
+#### GET
+
+- no options: original GET operation, will get the master version
+- `versionId: <versionId>` retrieve a specific version
+
+The implementation of a GET operation does not change compared to the standard
+version. A standard GET without versioning information would get the master
+version of a key. A version-specific GET would retrieve the specific version
+identified by the key for that version.
+
+#### LIST
+
+For a standard LIST on a bucket, metadata iterates through the keys by using
+the separator represented by `.` as an extra delimiter. For a listing of all
+versions of a bucket, there is no change compared to the original listing
+function. Instead, the API component returns all the keys in a List Objects call
+and filters for just the keys of the master versions in a List Object Versions
+call.
+
+For example, a standard LIST operation against the keys in a table below would
+return from metadata the list of `[ foo/bar, bar, qux/quz, quz ]` using the
+separator represented by `.` as the extra delimiter.
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar.v2 |
+| foo/bar.v1 |
+| bar |
+| qux/quz |
+| qux/quz.v2 |
+| qux/quz.v1 |
+| quz |
+| quz.v2 |
+| quz.v1 |
+
+## Implementation of Bucket Versioning in API
+
+### Object Metadata Versioning Attributes
+
+To access all the information needed to properly handle all cases that may exist
+in versioned operations, the API stores certain versioning-related information
+in the metadata attributes of each version's object metadata.
+
+These are the versioning-related metadata properties:
+
+- `isNull`: whether the version being stored is a null version.
+- `nullVersionId`: the version ID of the latest null version.
```

Comment:
I updated the description (was inaccurate as it was), and explained further in the section "Overwriting a Null Version That is Not Latest Version"